torch/autograd/gradcheck.py

```diff
@@ -19,21 +19,39 @@ def zero_gradients(x):
             zero_gradients(elem)
 
 
-def make_jacobian(input, num_out):
-    if is_tensor_like(input):
-        if not input.is_floating_point() and not input.is_complex():
-            return None
-        if not input.requires_grad:
-            return None
-        return input.new_zeros((input.nelement(), num_out), dtype=input.dtype, layout=torch.strided)
-    elif isinstance(input, collections.abc.Iterable) and not isinstance(input, str):
-        jacobians = list(filter(
-            lambda x: x is not None, (make_jacobian(elem, num_out) for elem in input)))
-        if not jacobians:
-            return None
-        return type(input)(jacobians)  # type: ignore
-    else:
-        return None
+def is_float_or_complex_tensor(obj):
+    return is_tensor_like(obj) and (obj.is_floating_point() or obj.is_complex())
+
+
+def make_jacobians_with_inputs(input_tensors: Tuple, dim=None):
+    # Makes zero-filled tensors from inputs. If `dim` is not None, for each tensor in
+    # `input_tensors`, returns a new zero-filled tensor with height of `t.numel` and width
+    # of `dim`. Otherwise, for each tensor, returns a 1-d tensor with size `(t.numel,)`.
+    # Each new tensor will be strided and have the same dtype and device as those of the
+    # corresponding input
+    out: List[torch.Tensor] = []
+    for t in input_tensors:
+        if is_float_or_complex_tensor(t) and t.requires_grad:
+            if dim is None:
+                out.append(t.new_zeros((t.nelement(),), layout=torch.strided))
+            else:
+                out.append(t.new_zeros((t.nelement(), dim), layout=torch.strided))
+    return tuple(out)
+
+
+def make_jacobians_with_outputs(output_tensors: Tuple, dtype=None, device=None, dim=None):
```

Comment:
Why have a `dim=None`? It looks like the only callsite uses `dim=input.numel()`. Ditto for `make_jacobians_with_inputs`.

Also nit: I always thought "make_jacobian" was a nondescriptive name -- we're not making the jacobian, we're just allocating memory for it.

Reply:
It had dim=None because I wanted to prepare for the fast gradcheck version, but I guess I should just introduce it there instead. make_jacobians_with_inputs turned out to not need the dim=None case at all though even in the fast case, so removing that entirely.

I'm updating it to be `allocate_jacobians_with_x`